Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp

```diff
@@ -0,0 +1,349 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "AudioDecoderCocoa.h"
+
+#if ENABLE(WEB_CODECS) && USE(AVFOUNDATION)
+
+#include "AudioSampleBufferConverter.h"
+#include "AudioSampleFormat.h"
+#include "CAAudioStreamDescription.h"
+#include "MediaSampleAVFObjC.h"
+#include "MediaUtilities.h"
+#include "PlatformRawAudioData.h"
+#include "SharedBuffer.h"
+#include <CoreAudio/CoreAudioTypes.h>
+#include <wtf/NeverDestroyed.h>
+#include <wtf/TZoneMallocInlines.h>
+#include <wtf/UniqueRef.h>
+#include <wtf/WorkQueue.h>
+#include <wtf/text/MakeString.h>
+
+#include <pal/cf/AudioToolboxSoftLink.h>
+#include <pal/cf/CoreMediaSoftLink.h>
+
+namespace WebCore {
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(AudioDecoderCocoa);
+
+static WorkQueue& queueSingleton()
+{
+    static std::once_flag onceKey;
+    static LazyNeverDestroyed<Ref<WorkQueue>> workQueue;
+    std::call_once(onceKey, [] {
+        workQueue.construct(WorkQueue::create("AudioDecoder queue"_s));
+    });
+    return workQueue.get();
+}
+
+class InternalAudioDecoderCocoa : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<InternalAudioDecoderCocoa> {
+    WTF_MAKE_TZONE_ALLOCATED_INLINE(InternalAudioDecoderCocoa);
+
+public:
+    static Ref<InternalAudioDecoderCocoa> create(AudioDecoder::OutputCallback&& outputCallback)
+    {
+        return adoptRef(*new InternalAudioDecoderCocoa(WTFMove(outputCallback)));
+    }
+    virtual ~InternalAudioDecoderCocoa() = default;
+
+    String initialize(const String& codecName, const AudioDecoder::Config&);
+
+    Ref<AudioDecoder::DecodePromise> decode(Ref<SharedBuffer>&&, bool isKeyFrame, int64_t timestamp, std::optional<uint64_t> duration);
+    Ref<GenericPromise> flush();
+    void close();
+    bool isConfigured() const { return !!m_converter; }
+
+private:
+    InternalAudioDecoderCocoa(AudioDecoder::OutputCallback&&);
+
+    static void decompressedAudioOutputBufferCallback(void*, CMBufferQueueTriggerToken);
+    Ref<AudioSampleBufferConverter> converter() { return *m_converter; }
+    void processedDecodedOutputs();
+
+    AudioDecoder::OutputCallback m_outputCallback;
+    RefPtr<AudioSampleBufferConverter> m_converter;
```

Comment:
Maybe we should have a static function to create the converter and pass it to InternalAudioDecoderCocoa constructor. Then we could make it a `const Ref`.